packages/pds/src/sql-repo-storage.ts

```diff
@@ -0,0 +1,200 @@
+import { RepoStorage } from '@atproto/repo'
+import { CID } from 'multiformats/cid'
+import Database from './db'
+import { IpldBlock } from './db/tables/ipld-block'
+import { IpldBlockCreator } from './db/tables/ipld-block-creator'
+import { RepoCommitBlock } from './db/tables/repo-commit-block'
+
+export class SqlRepoStorage extends RepoStorage {
+  constructor(
+    public db: Database,
+    public did: string,
+    public timestamp?: string,
+  ) {
+    super()
+  }
+
+  async getHead(forUpdate?: boolean): Promise<CID | null> {
+    let builder = this.db.db
+      .selectFrom('repo_root')
+      .selectAll()
+      .where('did', '=', this.did)
+    if (forUpdate) {
+      this.db.assertTransaction()
+      if (this.db.dialect !== 'sqlite') {
+        // SELECT FOR UPDATE is not supported by sqlite, but sqlite txs are SERIALIZABLE so we don't actually need it
+        builder = builder.forUpdate()
+      }
+    }
+    const found = await builder.executeTakeFirst()
+    return found ? CID.parse(found.root) : null
+  }
+
+  async getSavedBytes(cid: CID): Promise<Uint8Array | null> {
+    const found = await this.db.db
+      .selectFrom('ipld_block')
+      .where('cid', '=', cid.toString())
+      .select('content')
+      .executeTakeFirst()
+    return found ? found.content : null
+  }
+
+  async hasSavedBlock(cid: CID): Promise<boolean> {
+    const found = await this.db.db
+      .selectFrom('ipld_block')
+      .where('cid', '=', cid.toString())
+      .select('cid')
+      .executeTakeFirst()
+    return !!found
+  }
+
+  async putBlock(cid: CID, block: Uint8Array): Promise<void> {
+    this.db.assertTransaction()
+    const insertBlock = this.db.db
+      .insertInto('ipld_block')
+      .values({
+        cid: cid.toString(),
+        size: block.length,
+        content: block,
+        indexedAt: this.timestamp || new Date().toISOString(),
+      })
+      .onConflict((oc) => oc.doNothing())
+      .execute()
+    const insertCreator = this.db.db
+      .insertInto('ipld_block_creator')
+      .values({
+        cid: cid.toString(),
+        did: this.did,
+      })
+      .onConflict((oc) => oc.doNothing())
+      .execute()
+    await Promise.all([insertBlock, insertCreator])
+  }
+
+  async putMany(toPut: Map<string, Uint8Array>): Promise<void> {
```

Comment:
This will hit a limit on the size of `toPut` based on the pg/sqlite max parameter counts.  If it's a concern, a common workaround is to perform batches of inserts.

Reply:
yeah good point. we shouldn't be hitting that ever but doesn't hurt to add in